internal/engine/build_and_deployer.go

```diff
@@ -50,13 +54,27 @@ func NewCompositeBuildAndDeployer(builders BuildOrder, shouldFallBack FallbackTe
 	}
 }
 
+func (composite *CompositeBuildAndDeployer) forgetImage(ctx context.Context, image reference.NamedTagged) error {
+	for _, builder := range composite.builders {
+		err := builder.forgetImage(ctx, image)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
 func (composite *CompositeBuildAndDeployer) BuildAndDeploy(ctx context.Context, manifest model.Manifest, currentState BuildState) (BuildResult, error) {
 	var lastErr error
 	for _, builder := range composite.builders {
 		br, err := builder.BuildAndDeploy(ctx, manifest, currentState)
 		if err == nil {
 			// TODO(maia): maybe this only needs to be called after certain builds?
 			// I.e. should be called after image build but not after a successful container build?
+			if currentState.LastImage() != nil && br.Image != currentState.LastImage() {
+				composite.forgetImage(ctx, currentState.LastImage())
```

Comment:
it seems weird that this is a separate api...? could we make this part of PostProcessBuild?

Reply:
good call

Reply:
++